### PR TITLE
 Add CMakeLists.txt 

### PR DESCRIPTION
**Pull request by Frago9876543210**

### CODE REVIEW

**Comment by kubo** [comment]
@Frago9876543210 Your first commit https://github.com/kubo/funchook/pull/8/commits/3a559a2d6c0c4cf14937ce96315582dc8e29a81c is almost same with @minecraft-linux's [this commit](https://github.com/minecraft-linux/funchook/commit/7f3b9cfac6d17a18810122d93c5a24a82645cc2b). Is it by accident?

As for the second commit https://github.com/kubo/funchook/pull/8/commits/a58d512ae36be3d79374f48c6ef665215f5b49cf, I'll fix it by either removing config.h or creating config.h using [configure_file](https://cmake.org/cmake/help/latest/command/configure_file.html).

**Comment by Frago9876543210** [comment]
I'm trying to compile test with cmake. x64_64 and x86 works fine, but if compile it using `x86_64-w64-mingw32-gcc` or `i686-w64-mingw32-gcc` got following:
```
-- Looking for __GNU_LIBRARY__
-- Looking for __GNU_LIBRARY__ - not found
-- Looking for _GNU_SOURCE
-- Looking for _GNU_SOURCE - not found
```

**Comment by Frago9876543210** [comment]
Well, now it works on Linux and Mac OS but I still don't know how to fix the problem mentioned above

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,36 @@
+cmake_minimum_required(VERSION 3.10)
+
+project(funchook C ASM)
+
+include(CheckSymbolExists)
+
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -O2")
```
Comment (kubo) [inline CMakeLists.txt]: Could you remove this line?
When cmake is executed with `-DCMAKE_BUILD_TYPE=Release`, `-O3` is added to CFLAGS.

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+if (NOT DEFINED _GNU_SOURCE)
+    check_symbol_exists(__GNU_LIBRARY__ "features.h" _GNU_SOURCE)
+
+    if (NOT _GNU_SOURCE)
+        unset(_GNU_SOURCE CACHE)
+        check_symbol_exists(_GNU_SOURCE "features.h" _GNU_SOURCE)
+    endif ()
+endif ()
```
Comment (kubo) [inline CMakeLists.txt]: Could you remove code related to `_GNU_SOURCE` in this file and add the following code just before `#include <stdio.h>` in funchook_unix.c as in src/funchook_io.c and src/os_func_unix.c？
```
#if defined __linux && !defined(_GNU_SOURCE)
#define _GNU_SOURCE
#endif
```

**Comment by kubo** [comment]
> I'm trying to compile test with cmake. x64_64 and x86 works fine, but if compile it using x86_64-w64-mingw32-gcc or i686-w64-mingw32-gcc got following:
> 
> ```
> ~/funchook/src/funchook_unix.c:37:10: fatal error: sys/mman.h: No such file or directory.
>  #include <sys/mman.h>
>           ^~~~~~~~~~~~
> ```

Mingw32 is a cross compilation tool to build programs for Windows. So `funchook_unix.c` is not available. Use `funchook_windows.c` instead. See [here](https://gitlab.kitware.com/cmake/community/wikis/doc/cmake/CrossCompiling#the-toolchain-file) and [here](https://gist.github.com/peterspackman/8cf73f7f12ba270aa8192d6911972fe8) to use mingw32 with cmake. 

May I merge and close this issue after you fix code pointed before?
Otherwise, may I wait mingw32 support?

**Comment by Frago9876543210** [comment]
I'll try to fix above issues and will add mingw32 support

**Comment by Frago9876543210** [comment]
Tests still not works. I got following:
```
$ cd test
$ mkdir -p build && cd build
$ cmake -DCMAKE_C_COMPILER=$(where x86_64-w64-mingw32-gcc) .. && make VERBOSE=1

-- The C compiler identification is GNU 7.3.0
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/x86_64-w64-mingw32-gcc
-- Check for working C compiler: /usr/bin/x86_64-w64-mingw32-gcc
-- Check for working C compiler: /usr/bin/x86_64-w64-mingw32-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /home/alex/funchook/test/build
/usr/bin/cmake -H/home/alex/funchook/test -B/home/alex/funchook/test/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /home/alex/funchook/test/build/CMakeFiles /home/alex/funchook/test/build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: вход в каталог «/home/alex/funchook/test/build»
make -f ../../build/CMakeFiles/distorm.dir/build.make ../../build/CMakeFiles/distorm.dir/depend
make[2]: вход в каталог «/home/alex/funchook/test/build»
cd /home/alex/funchook/test/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/alex/funchook/test /home/alex/funchook /home/alex/funchook/test/build /home/alex/funchook/build /home/alex/funchook/build/CMakeFiles/distorm.dir/DependInfo.cmake --color=
Dependee "/home/alex/funchook/build/CMakeFiles/distorm.dir/DependInfo.cmake" is newer than depender "/home/alex/funchook/build/CMakeFiles/distorm.dir/depend.internal".
Dependee "/home/alex/funchook/build/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/home/alex/funchook/build/CMakeFiles/distorm.dir/depend.internal".
Scanning dependencies of target distorm
make[2]: выход из каталога «/home/alex/funchook/test/build»
make -f ../../build/CMakeFiles/distorm.dir/build.make ../../build/CMakeFiles/distorm.dir/build
make[2]: вход в каталог «/home/alex/funchook/test/build»
[  4%] Building C object ../../build/CMakeFiles/distorm.dir/distorm/src/decoder.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/distorm/include   -o CMakeFiles/distorm.dir/distorm/src/decoder.c.o   -c /home/alex/funchook/distorm/src/decoder.c
[  8%] Building C object ../../build/CMakeFiles/distorm.dir/distorm/src/distorm.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/distorm/include   -o CMakeFiles/distorm.dir/distorm/src/distorm.c.o   -c /home/alex/funchook/distorm/src/distorm.c
[ 12%] Building C object ../../build/CMakeFiles/distorm.dir/distorm/src/instructions.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/distorm/include   -o CMakeFiles/distorm.dir/distorm/src/instructions.c.o   -c /home/alex/funchook/distorm/src/instructions.c
[ 16%] Building C object ../../build/CMakeFiles/distorm.dir/distorm/src/insts.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/distorm/include   -o CMakeFiles/distorm.dir/distorm/src/insts.c.o   -c /home/alex/funchook/distorm/src/insts.c
[ 20%] Building C object ../../build/CMakeFiles/distorm.dir/distorm/src/mnemonics.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/distorm/include   -o CMakeFiles/distorm.dir/distorm/src/mnemonics.c.o   -c /home/alex/funchook/distorm/src/mnemonics.c
[ 25%] Building C object ../../build/CMakeFiles/distorm.dir/distorm/src/operands.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/distorm/include   -o CMakeFiles/distorm.dir/distorm/src/operands.c.o   -c /home/alex/funchook/distorm/src/operands.c
[ 29%] Building C object ../../build/CMakeFiles/distorm.dir/distorm/src/prefix.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/distorm/include   -o CMakeFiles/distorm.dir/distorm/src/prefix.c.o   -c /home/alex/funchook/distorm/src/prefix.c
[ 33%] Building C object ../../build/CMakeFiles/distorm.dir/distorm/src/textdefs.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/distorm/include   -o CMakeFiles/distorm.dir/distorm/src/textdefs.c.o   -c /home/alex/funchook/distorm/src/textdefs.c
[ 37%] Building C object ../../build/CMakeFiles/distorm.dir/distorm/src/wstring.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/distorm/include   -o CMakeFiles/distorm.dir/distorm/src/wstring.c.o   -c /home/alex/funchook/distorm/src/wstring.c
[ 41%] Linking C static library libdistorm.a
cd /home/alex/funchook/build && /usr/bin/cmake -P CMakeFiles/distorm.dir/cmake_clean_target.cmake
cd /home/alex/funchook/build && /usr/bin/cmake -E cmake_link_script CMakeFiles/distorm.dir/link.txt --verbose=1
/usr/bin/ar qc libdistorm.a  CMakeFiles/distorm.dir/distorm/src/decoder.c.o CMakeFiles/distorm.dir/distorm/src/distorm.c.o CMakeFiles/distorm.dir/distorm/src/instructions.c.o CMakeFiles/distorm.dir/distorm/src/insts.c.o CMakeFiles/distorm.dir/distorm/src/mnemonics.c.o CMakeFiles/distorm.dir/distorm/src/operands.c.o CMakeFiles/distorm.dir/distorm/src/prefix.c.o CMakeFiles/distorm.dir/distorm/src/textdefs.c.o CMakeFiles/distorm.dir/distorm/src/wstring.c.o
/usr/bin/ranlib libdistorm.a
make[2]: выход из каталога «/home/alex/funchook/test/build»
[ 41%] Built target distorm
make -f ../../build/CMakeFiles/funchook.dir/build.make ../../build/CMakeFiles/funchook.dir/depend
make[2]: вход в каталог «/home/alex/funchook/test/build»
cd /home/alex/funchook/test/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/alex/funchook/test /home/alex/funchook /home/alex/funchook/test/build /home/alex/funchook/build /home/alex/funchook/build/CMakeFiles/funchook.dir/DependInfo.cmake --color=
Dependee "/home/alex/funchook/build/CMakeFiles/funchook.dir/DependInfo.cmake" is newer than depender "/home/alex/funchook/build/CMakeFiles/funchook.dir/depend.internal".
Dependee "/home/alex/funchook/build/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/home/alex/funchook/build/CMakeFiles/funchook.dir/depend.internal".
Scanning dependencies of target funchook
make[2]: выход из каталога «/home/alex/funchook/test/build»
make -f ../../build/CMakeFiles/funchook.dir/build.make ../../build/CMakeFiles/funchook.dir/build
make[2]: вход в каталог «/home/alex/funchook/test/build»
[ 45%] Building C object ../../build/CMakeFiles/funchook.dir/src/funchook.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/include -I/home/alex/funchook/distorm/include   -o CMakeFiles/funchook.dir/src/funchook.c.o   -c /home/alex/funchook/src/funchook.c
/home/alex/funchook/src/funchook.c:120:13: warning: ‘funchook_create’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 funchook_t *funchook_create(void)
             ^~~~~~~~~~~~~~~
/home/alex/funchook/src/funchook.c:130:5: warning: ‘funchook_prepare’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 int funchook_prepare(funchook_t *funchook, void **target_func, void *hook_func)
     ^~~~~~~~~~~~~~~~
/home/alex/funchook/src/funchook.c:142:5: warning: ‘funchook_install’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 int funchook_install(funchook_t *funchook, int flags)
     ^~~~~~~~~~~~~~~~
/home/alex/funchook/src/funchook.c:152:5: warning: ‘funchook_uninstall’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 int funchook_uninstall(funchook_t *funchook, int flags)
     ^~~~~~~~~~~~~~~~~~
/home/alex/funchook/src/funchook.c:162:5: warning: ‘funchook_destroy’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 int funchook_destroy(funchook_t *funchook)
     ^~~~~~~~~~~~~~~~
/home/alex/funchook/src/funchook.c:172:13: warning: ‘funchook_error_message’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 const char *funchook_error_message(const funchook_t *funchook)
             ^~~~~~~~~~~~~~~~~~~~~~
/home/alex/funchook/src/funchook.c:177:5: warning: ‘funchook_set_debug_file’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 int funchook_set_debug_file(const char *name)
     ^~~~~~~~~~~~~~~~~~~~~~~
[ 50%] Building C object ../../build/CMakeFiles/funchook.dir/src/funchook_io.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/include -I/home/alex/funchook/distorm/include   -o CMakeFiles/funchook.dir/src/funchook_io.c.o   -c /home/alex/funchook/src/funchook_io.c
[ 54%] Building C object ../../build/CMakeFiles/funchook.dir/src/funchook_x86.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/include -I/home/alex/funchook/distorm/include   -o CMakeFiles/funchook.dir/src/funchook_x86.c.o   -c /home/alex/funchook/src/funchook_x86.c
[ 58%] Building C object ../../build/CMakeFiles/funchook.dir/src/os_func.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/include -I/home/alex/funchook/distorm/include   -o CMakeFiles/funchook.dir/src/os_func.c.o   -c /home/alex/funchook/src/os_func.c
[ 62%] Building C object ../../build/CMakeFiles/funchook.dir/src/printf_base.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/include -I/home/alex/funchook/distorm/include   -o CMakeFiles/funchook.dir/src/printf_base.c.o   -c /home/alex/funchook/src/printf_base.c
[ 66%] Building C object ../../build/CMakeFiles/funchook.dir/src/funchook_windows.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/include -I/home/alex/funchook/distorm/include   -o CMakeFiles/funchook.dir/src/funchook_windows.c.o   -c /home/alex/funchook/src/funchook_windows.c
[ 70%] Building C object ../../build/CMakeFiles/funchook.dir/src/os_func_windows.c.o
cd /home/alex/funchook/build && /usr/bin/x86_64-w64-mingw32-gcc  -I/home/alex/funchook/include -I/home/alex/funchook/distorm/include   -o CMakeFiles/funchook.dir/src/os_func_windows.c.o   -c /home/alex/funchook/src/os_func_windows.c
[ 75%] Linking C static library libfunchook.a
cd /home/alex/funchook/build && /usr/bin/cmake -P CMakeFiles/funchook.dir/cmake_clean_target.cmake
cd /home/alex/funchook/build && /usr/bin/cmake -E cmake_link_script CMakeFiles/funchook.dir/link.txt --verbose=1
/usr/bin/ar qc libfunchook.a  CMakeFiles/funchook.dir/src/funchook.c.o CMakeFiles/funchook.dir/src/funchook_io.c.o CMakeFiles/funchook.dir/src/funchook_x86.c.o CMakeFiles/funchook.dir/src/os_func.c.o CMakeFiles/funchook.dir/src/printf_base.c.o CMakeFiles/funchook.dir/src/funchook_windows.c.o CMakeFiles/funchook.dir/src/os_func_windows.c.o
/usr/bin/ranlib libfunchook.a
make[2]: выход из каталога «/home/alex/funchook/test/build»
[ 75%] Built target funchook
make -f CMakeFiles/funchook_test.dir/build.make CMakeFiles/funchook_test.dir/depend
make[2]: вход в каталог «/home/alex/funchook/test/build»
cd /home/alex/funchook/test/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/alex/funchook/test /home/alex/funchook/test /home/alex/funchook/test/build /home/alex/funchook/test/build /home/alex/funchook/test/build/CMakeFiles/funchook_test.dir/DependInfo.cmake --color=
Dependee "/home/alex/funchook/test/build/CMakeFiles/funchook_test.dir/DependInfo.cmake" is newer than depender "/home/alex/funchook/test/build/CMakeFiles/funchook_test.dir/depend.internal".
Dependee "/home/alex/funchook/test/build/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/home/alex/funchook/test/build/CMakeFiles/funchook_test.dir/depend.internal".
Scanning dependencies of target funchook_test
make[2]: выход из каталога «/home/alex/funchook/test/build»
make -f CMakeFiles/funchook_test.dir/build.make CMakeFiles/funchook_test.dir/build
make[2]: вход в каталог «/home/alex/funchook/test/build»
[ 79%] Building C object CMakeFiles/funchook_test.dir/libfunchook_test.c.o
/usr/bin/x86_64-w64-mingw32-gcc -Dfunchook_test_EXPORTS -I/home/alex/funchook/include -I/home/alex/funchook/distorm/include  -fPIC   -o CMakeFiles/funchook_test.dir/libfunchook_test.c.o   -c /home/alex/funchook/test/libfunchook_test.c
[ 83%] Building C object CMakeFiles/funchook_test.dir/libfunchook_test2.c.o
/usr/bin/x86_64-w64-mingw32-gcc -Dfunchook_test_EXPORTS -I/home/alex/funchook/include -I/home/alex/funchook/distorm/include  -fPIC   -o CMakeFiles/funchook_test.dir/libfunchook_test2.c.o   -c /home/alex/funchook/test/libfunchook_test2.c
[ 87%] Linking C shared library libfunchook_test.so
/usr/bin/cmake -E cmake_link_script CMakeFiles/funchook_test.dir/link.txt --verbose=1
/usr/bin/x86_64-w64-mingw32-gcc -fPIC   -shared -Wl,-soname,libfunchook_test.so -o libfunchook_test.so CMakeFiles/funchook_test.dir/libfunchook_test.c.o CMakeFiles/funchook_test.dir/libfunchook_test2.c.o /home/alex/funchook/build/libfunchook.a /home/alex/funchook/build/libdistorm.a -lpsapi 
CMakeFiles/funchook_test.dir/libfunchook_test2.c.o:libfunchook_test2.c:(.text+0x9): undefined reference to `get_val_in_exe'
collect2: error: ld returned 1 exit status
CMakeFiles/funchook_test.dir/build.make:122: recipe for target 'libfunchook_test.so' failed
make[2]: *** [libfunchook_test.so] Error 1
make[2]: выход из каталога «/home/alex/funchook/test/build»
CMakeFiles/Makefile2:68: recipe for target 'CMakeFiles/funchook_test.dir/all' failed
make[1]: *** [CMakeFiles/funchook_test.dir/all] Error 2
make[1]: выход из каталога «/home/alex/funchook/test/build»
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by kubo** [comment]
```
/home/alex/funchook/src/funchook.c:120:13: warning: ‘funchook_create’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
 funchook_t *funchook_create(void)
             ^~~~~~~~~~~~~~~
```
Could you define `FUNCHOOK_EXPORTS` to use `__declspec(dllexport)`? When a function is defined without `static` in a shared library, it is visible from the outside of the shared library on Unix. On the other hand, it is invisible on Windows without [`__declspec(dllexport)`](https://docs.microsoft.com/en-us/cpp/build/exporting-from-a-dll-using-declspec-dllexport?view=vs-2019) or a [DEF file](https://docs.microsoft.com/en-us/cpp/build/exporting-from-a-dll-using-def-files?view=vs-2019).
https://github.com/kubo/funchook/blob/a61a080438979e05a7c8b125e0e0f37b978947e0/include/funchook.h#L38-L42

```
/usr/bin/x86_64-w64-mingw32-gcc -fPIC   -shared -Wl,-soname,libfunchook_test.so -o libfunchook_test.so CMakeFiles/funchook_test.dir/libfunchook_test.c.o CMakeFiles/funchook_test.dir/libfunchook_test2.c.o /home/alex/funchook/build/libfunchook.a /home/alex/funchook/build/libdistorm.a -lpsapi 
CMakeFiles/funchook_test.dir/libfunchook_test2.c.o:libfunchook_test2.c:(.text+0x9): undefined reference to `get_val_in_exe'
```
`libfunchook_test.so` refers `get_val_in_exe` defined in `funchook_test_ex`. When the symbol is not resolved at compile time, it is resolved at runtime on Unix. On the other hand, it must be resolved at compile time on Windows. The following code in `Makefile.in` makes an import library to tell that `get_val_in_exe` is in `funchook_test.exe` on Windows.
https://github.com/kubo/funchook/blob/a61a080438979e05a7c8b125e0e0f37b978947e0/test/Makefile.in#L39-L43
And the following two pieces passes the import library to the linker.
https://github.com/kubo/funchook/blob/a61a080438979e05a7c8b125e0e0f37b978947e0/test/Makefile.in#L16
https://github.com/kubo/funchook/blob/a61a080438979e05a7c8b125e0e0f37b978947e0/test/Makefile.in#L30-L31

**Comment by Frago9876543210** [comment]
I don't understand how to get .lib file using cmake like is it done here
```
x86_64-w64-mingw32-gcc -shared -Wl,--out-implib,funchook.lib  -o funchook.dll os_func.o os_func_windows.o printf_base.o funchook.o funchook_io.o funchook_x86.o funchook_windows.o mnemonics.o wstring.o textdefs.o prefix.o operands.o insts.o instructions.o distorm.o decoder.o  -lpsapi
```

**Comment by kubo** [comment]
> I don't understand how to get .lib file using cmake like is it done here

It is created by cmake when a shared library is created.
```shell
$ cat CMakeLists.txt  # Creates a shared library
cmake_minimum_required(VERSION 3.5)

project(foo C)
add_library(foo SHARED foo.c)
$ cat foo.c  # Source code of the shared library
int add_one(int a)
{
    return a + 1;
}
$ cat mingw-w64-x86_64.cmake  # Toolchain file for cross compilation using mingw32
set(CMAKE_SYSTEM_NAME Windows)
set(TOOLCHAIN_PREFIX x86_64-w64-mingw32)

# cross compilers to use for C, C++ and Fortran
set(CMAKE_C_COMPILER ${TOOLCHAIN_PREFIX}-gcc)
set(CMAKE_CXX_COMPILER ${TOOLCHAIN_PREFIX}-g++)
set(CMAKE_Fortran_COMPILER ${TOOLCHAIN_PREFIX}-gfortran)
set(CMAKE_RC_COMPILER ${TOOLCHAIN_PREFIX}-windres)

# target environment on the build host system
set(CMAKE_FIND_ROOT_PATH /usr/${TOOLCHAIN_PREFIX})

# modify default behavior of FIND_XXX() commands
set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
$ mkdir build && cd build
$ cmake .. -DCMAKE_TOOLCHAIN_FILE=mingw-w64-x86_64.cmake
-- The C compiler identification is GNU 5.3.1
-- Check for working C compiler: /usr/bin/x86_64-w64-mingw32-gcc
-- Check for working C compiler: /usr/bin/x86_64-w64-mingw32-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /home/kubo/foo/build
$ make
[ 50%] Building C object CMakeFiles/foo.dir/foo.c.obj
[100%] Linking C shared library libfoo.dll
[100%] Built target foo
$ ls -l
total 140
-rw-rw-r-- 1 kubo kubo  10936 Jul 24 07:20 CMakeCache.txt
drwxrwxr-x 5 kubo kubo   4096 Jul 24 07:23 CMakeFiles
-rw-rw-r-- 1 kubo kubo   4605 Jul 24 07:20 Makefile
-rw-rw-r-- 1 kubo kubo   1206 Jul 24 07:20 cmake_install.cmake
-rwxrwxr-x 1 kubo kubo 106627 Jul 24 07:23 libfoo.dll
-rw-rw-r-- 1 kubo kubo   1384 Jul 24 07:23 libfoo.dll.a
```
`libfoo.dll` and `libfoo.dll.a` were created. The latter is the import library for the former.

**Comment by Frago9876543210** [comment]
I'm not sure about tests. `funchook_test_ex` compiles, but only if `funchook_test` __STATIC__ 

<details>
  <summary>mingw-w64-x86_64.cmake</summary>

```
set(CMAKE_SYSTEM_NAME Windows)
set(TOOLCHAIN_PREFIX x86_64-w64-mingw32)

# cross compilers to use for C, C++ and Fortran
set(CMAKE_C_COMPILER ${TOOLCHAIN_PREFIX}-gcc)
set(CMAKE_CXX_COMPILER ${TOOLCHAIN_PREFIX}-g++)
set(CMAKE_Fortran_COMPILER ${TOOLCHAIN_PREFIX}-gfortran)
set(CMAKE_RC_COMPILER ${TOOLCHAIN_PREFIX}-windres)

if (NOT DLLTOOL)
    set(DLLTOOL ${TOOLCHAIN_PREFIX}-dlltool)
endif ()

# target environment on the build host system
set(CMAKE_FIND_ROOT_PATH /usr/${TOOLCHAIN_PREFIX})

# modify default behavior of FIND_XXX() commands
set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
```
</details>

<details>
  <summary>test/CMakeLists.txt</summary>

```
cmake_minimum_required(VERSION 3.10)
project(libfunchook_test C ASM)

if (${CMAKE_C_FLAGS} MATCHES "-m32")
    set(CMAKE_ASM_FLAGS "-m32")
    set(FUNCHOOK_TEST "x86_test.S")
else ()
    set(FUNCHOOK_TEST "x86_64_test.S")
endif ()

add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/../ ${CMAKE_CURRENT_SOURCE_DIR}/../build)

add_executable(funchook_test_ex test_main.c ${FUNCHOOK_TEST})
add_custom_command(TARGET funchook_test_ex PRE_BUILD
        WORKING_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}"
        COMMAND echo "LIBRARY funchook_test.exe" > funchook_test_exe.def
        COMMAND echo "EXPORTS" >> funchook_test_exe.def
        COMMAND echo "get_val_in_exe" >> funchook_test_exe.def
        COMMAND ${DLLTOOL} -d funchook_test_exe.def -l funchook_test_exe.lib)
target_link_libraries(funchook_test_ex funchook m funchook_test)

add_library(funchook_test STATIC libfunchook_test.c libfunchook_test2.c)
if (APPLE)
    set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -undefined dynamic_lookup")
endif ()
target_link_libraries(funchook_test funchook m)
```
</details>

**Comment by kubo** [comment]
> I'm not sure about tests. funchook_test_ex compiles, but only if funchook_test **STATIC**

It must be a shared library to check whether functions themselves, not "jump table", are hooked. See the second method in [this section](http://www.reverse-engineering.info/SystemInformation/iat.html#sec7) about "jump table".

When the third argument of funchook_prepare points to a jump table, funchook uses the address of corresponding external function instead.

**Pull request closed without merging**